### PR TITLE
Restaurant form

### PR DESCRIPTION
**Pull request by ZeshanShahid79**

### CODE REVIEW

File: src/components/ContactForm/ContactForm.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+export default function ContactForm() {
+  return (
+    <StyledContactForm aria-labelledby="formHeader" onSubmit={onSubmit}>
+      <h2 id="formHeader">Contact Us</h2>
+      <Input
+        labelText="Restaurantname"
+        id="restaurantname"
+        name="restaurantname"
+        type="text"
+        required={true}
```
Comment (Sven-Lyco) [inline src/components/ContactForm/ContactForm.js]: I'm not sure, but I think you could set the required without `true`. Try it out for your input fields
```js
required
```

File: src/components/ContactForm/ContactForm.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import styled from 'styled-components';
+import Input from '../Input/Input';
+import emailjs from 'emailjs-com';
+
+export default function ContactForm() {
+  return (
+    <StyledContactForm aria-labelledby="formHeader" onSubmit={onSubmit}>
```
Comment (Sven-Lyco) [inline src/components/ContactForm/ContactForm.js]: You should rename the function inside the curlybrackets like this:
```js
onSubmit={handleSubmit}
```

File: src/components/ContactForm/ContactForm.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        type="number"
+        required={true}
+      />
+      <label htmlFor="message">Message</label>
+      <textarea id="message" name="message" rows="4" />
+      <input data-testid="message" type="submit" value="Send" />
+    </StyledContactForm>
+  );
+
+  function onSubmit(event) {
```
Comment (Sven-Lyco) [inline src/components/ContactForm/ContactForm.js]: The function should have the name `handleSubmit`

File: src/components/ContactForm/ContactForm.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      <Input
+        labelText="Telephone"
+        id="telephone"
+        name="telephone"
+        type="number"
+        required={true}
+      />
+      <label htmlFor="message">Message</label>
+      <textarea id="message" name="message" rows="4" />
+      <input data-testid="message" type="submit" value="Send" />
```
Comment (Sven-Lyco) [inline src/components/ContactForm/ContactForm.js]: maybe you can take a button here?

File: src/components/ContactForm/ContactForm.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    event.preventDefault();
+    emailjs
+      .sendForm(
+        'service_hgexx04',
+        'template_db85ugq',
+        event.target,
+        'SHU7y_FFmZng3tHx3'
+      )
+      .then(res => {
+        console.log(res);
```
Comment (Sven-Lyco) [inline src/components/ContactForm/ContactForm.js]: Do you need this console.log?
I do not know what comes there as a response, but maybe you can display the message in an alert? :)

**Review by Sven-Lyco — COMMENTED** [review]
You did a good job here, just a few comments. Keep going 🚀

File: src/components/ContactForm/FormInput.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    display: block;
+  }
+`;
+
+const StyledSpan = styled.span`
+  font-size: 12px;
+  padding: 3px;
+  color: red;
+  display: none;
+`;
```
Comment (macfrei) [inline src/components/ContactForm/FormInput.js]: I would get the value as prop and then only render the span if the value is given. Try if this works.
```suggestion
  const { label, errorMessage, onChange, id, value ...inputProps } = props;

  return (
    <>
      <label>{label}</label>
      <StyledInput
        {...inputProps}
        onChange={onChange}
        onBlur={handleFocus}
        focused={focused.toString()}
        value={value}
      />
      {value.length === 0 && <StyledSpan>{errorMessage}</StyledSpan>}
    </>
  );
  function handleFocus() {
    setFocused(true);
  }
}

const StyledInput = styled.input`
  padding: 15px;
  margin: 10px 0px;
`;

const StyledSpan = styled.span`
  font-size: 12px;
  padding: 3px;
  color: red;
  display: block;
`;
```

File: src/components/FormInput/FormInput.spec.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { render, screen } from '@testing-library/react';
+import FormInput from './FormInput';
+
+describe('FormInput', () => {
+  it('renders a label and input with Name of input and a span with a conditional errormesasage', () => {
+    render(<FormInput label="Restaurantname" />);
```
Comment (macfrei) [inline src/components/FormInput/FormInput.spec.js]: You need to add all props here. `value` is now undefined, that's why your test is not working.
All of your props need to be passe:
```js
    {
      id: 2,
      name: 'address',
      type: 'text',
      errorMessage: 'Address should include Street and Number',
      label: 'Address',
      required: true,
      pattern: /[^A-Za-z\- . [0-9]$]/g,
    }
```

File: src/App.js [hunk 8]
```diff
@@ -1,17 +1,30 @@
...
+
+      <ContactForm />
     </>
   );
   function refreshPage() {
     window.location.reload(false);
   }
 }
+
+const RandomRestaurantPage = styled.main`
```
Comment (macfrei) [inline src/App.js:22]: I think the `ContactForm` should also be within `main`.

File: src/components/ContactForm/ContactForm.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    {
+      id: 4,
+      name: 'telephone',
+      type: 'text',
+      errorMessage: 'Telephonenumber should include only Numbers',
+      label: 'Telephone',
+      required: true,
+      pattern: /[^0-9]/g,
+    },
+  ];
```
Comment (macfrei) [inline src/components/ContactForm/ContactForm.js]: This is always the same. I would recommend to save this array outside the component call.

File: src/components/ContactForm/ContactForm.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    },
+    {
+      id: 3,
+      name: 'email',
+      type: 'email',
+      errorMessage: 'It should be a valid email address',
+      label: 'E-Mail',
+      required: true,
+      // pattern:
+      //   /^[a-zA-Z0-9][-.+!#$%&'*/=?^`{|]{0,1}([a-zA-Z0-9][-.+!#$%&'*/=?^`{|]{0,1})[a-zA-Z0-9]@[a-zA-Z0-9][-.]{0,1}([a-zA-Z][-.]{0,1})[a-zA-Z0-9].[a-zA-Z0-9]{1,}([.-]{0,1}[a-zA-Z]){0,}[a-zA-Z0-9]{0,}$/i,
```
Comment (macfrei) [inline src/components/ContactForm/ContactForm.js]: If you don't use something, please delete it and delete comments. We always try to have the `main` branch as clean as possible.

File: src/components/ContactForm/ContactForm.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
+import { useState } from 'react';
+import Button from '../Button/Button';
+import FormInput from '../FormInput/FormInput';
+import styled from 'styled-components';
+import emailjs from 'emailjs-com';
+
+export default function ContactForm2() {
```
Comment (macfrei) [inline src/components/ContactForm/ContactForm.js]:
```suggestion
export default function ContactForm() {
```
Why the `2`?

File: src/components/ContactForm/ContactForm.js [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    event.preventDefault();
+    emailjs
+      .sendForm(
+        'service_hgexx04',
+        'template_db85ugq',
+        event.target,
+        'SHU7y_FFmZng3tHx3'
+      )
+
+      .catch(err => console.log(err));
```
Comment (macfrei) [inline src/components/ContactForm/ContactForm.js:84]: Why are you sending an email? What is the use case? 🤔

File: src/components/FormInput/FormInput.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    props;
+
+  return (
+    <section aria-labelledby="inputlabel">
+      <label id="inputlabel">{label}</label>
+      <StyledInput
+        {...inputProps}
+        onChange={onChange}
+        onBlur={handleFocus}
+        focused={focused.toString()}
```
Comment (macfrei) [inline src/components/FormInput/FormInput.js]: Why are you doing this?

File: src/components/FormInput/FormInput.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        value={value}
+      />
```
Comment (macfrei) [inline src/components/FormInput/FormInput.js]: The label and input element are not connected. I would recommend to connect them as you usually do with label and input elements :)
```suggestion
      <label htmlFor={name}>{label}</label>
      <StyledInput
         id={name}
        {...inputProps}
        onChange={onChange}
        onBlur={handleFocus}
        focused={focused.toString()}
        value={value}
      />
```
For this to work you would need to destructure `name` as well from your props.

File: src/components/FormInput/FormInput.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
+import { useState } from 'react';
+import styled from 'styled-components';
+
+export default function FormInput(props) {
+  const [focused, setFocused] = useState(false);
+  const { label, errorMessage, onChange, id, pattern, value, ...inputProps } =
+    props;
+
+  return (
+    <section aria-labelledby="inputlabel">
```
Comment (macfrei) [inline src/components/FormInput/FormInput.js]: This is unnecessary if you connect the label/input correctly I think :)
```suggestion
    <section>
```

File: src/components/FormInput/FormInput.spec.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
+import { render, screen } from '@testing-library/react';
+import FormInput from './FormInput';
+
+describe('FormInput', () => {
+  it('renders a label and input with Name of input and a span with a conditional errormesasage', () => {
```
Comment (macfrei) [inline src/components/FormInput/FormInput.spec.js]:
```suggestion
  it('renders a label and input with Name of input and a span with a conditional error message', () => {
```

**Review by macfrei — COMMENTED** [review]
Nice job, I left some comments you could work though though 🙂. Just have one general question about sending the email and what the use case is etc.

**Pull request merged**